### PR TITLE
Fix part of #5025: App and OS Deprecation Milestone 4 - Gate the new Deprecation Dialogs and Add Logic to Display Them

### PR DESCRIPTION
**Pull request by kkmurerwa**
## Explanation

Fix part of #5025 - When this PR is merged, it will;
 - Add ability to display the new deprecation dialogs when the app-and-os-deprecation feature flag is enabled.
 - Add ability to cache previous responses so that the user is not spammed with dialogs every time they open the app.
 - Include tests for all added functionalities.

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

**Comment by kkmurerwa** [comment]
Hi @adhiamboperes. This PR is ready for review. PTAL

File: app/src/main/java/org/oppia/android/app/notice/DeprecationNoticeActionListener.kt
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/notice/DeprecationNoticeActionListener.kt]: Replace `option` with `button`.

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt]: You've switched these two actions.

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt]:
```suggestion
  /** Handles cases where the user clicks the update button on a deprecation notice dialog. */
```

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt]: Is this a replacement for `DeprecationNoticeExitAppListener` that is implemented by `SplashActivity`? I see you have both but it seems redundant

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt:241]: Wrapping the whole block dependent on the  feature flag value would likely mean that this block is not executed unless the feature flag is enabled/true.

an option is to switch to a `when` block without a variable, or do a seperate `when` block for when the `enableAppAndOsDeprecation` is true

File: domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt
Comment (adhiamboperes) [inline domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt]: Could you instead specify what the **old** logic is?

File: domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt
Comment (adhiamboperes) [inline domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt]: This can be extracted to a public api in the `DeprecationController`

File: domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt
Comment (adhiamboperes) [inline domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt]: We typically do not use **runBlocking** in production code. Dataproviders have functions for transforming the data, and you could do this in a suspend function instead. The ultimate choice will depend on the refactor per my suugestion below.

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
Thanks @kkmurerwa!. And apologies for the delay in review.

I have left comments, I think this needs two more review passes before it is  merge ready.

File: domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt
Comment (adhiamboperes) [reply domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt]: A simple controller handling data provider transformations: https://github.com/oppia/oppia-android/blob/develop/domain/src/main/java/org/oppia/android/domain/survey/SurveyGatingController.kt

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (kkmurerwa) [reply app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt]: I have both so that the old flow still works. I had removed the older `DeprecationNoticeExitAppListener` and Ben was of the idea that I should keep both until the feature is stable and then we can remove the older listener and all the code related to it. In the meantime, we should keep both implementations and only gate the feature.

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (kkmurerwa) [reply app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt:241]: I have a seperate block for the legacy deprecation feature. I have also moved the logic for the new deprecation feature to a seperate function so that it's clearer now.

File: domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt
Comment (kkmurerwa) [reply domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt]: I have migrated this logic to the specified class.

File: domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt
Comment (kkmurerwa) [reply domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt]: I have applied a transform and the code seems to work as it should now.

File: domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt
Comment (kkmurerwa) [reply domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt]: I have defined this better. Let me know if it is up-to-standard or if it requires some extra modification.

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (kkmurerwa) [reply app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt]: I have modified this instance.

File: app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
Comment (kkmurerwa) [reply app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt]: Thank you for spotting this. I have made the necessary changes.

File: app/src/main/java/org/oppia/android/app/notice/DeprecationNoticeActionListener.kt
Comment (kkmurerwa) [reply app/src/main/java/org/oppia/android/app/notice/DeprecationNoticeActionListener.kt]: I have replaced the three instances of `option` with `button`.

File: domain/src/main/java/org/oppia/android/domain/onboarding/DeprecationController.kt
Comment (adhiamboperes) [inline domain/src/main/java/org/oppia/android/domain/onboarding/DeprecationController.kt]: Leverage DataProviders
```suggestion
// call this in the `AppStartupStateController`
    fun getDeprecationResponse(): DataProvider<DeprecationResponse> {
    return deprecationStore.transform(GET_DEPRECATION_RESPONSE_DATABASE_ID) { deprecationResponseDb ->
      DeprecationResponse.newBuilder()
        .setDeprecationNoticeType(getDeprecationNoticeType(deprecationResponseDb))
        .build()
      //add the other fields
    }
  }

  private fun getDeprecationNoticeType(deprecationResponseDb: DeprecationResponseDatabase): DeprecationNoticeType {
    return if (deprecationResponseDb.hasAppDeprecationResponse()) {
      DeprecationNoticeType.APP_DEPRECATION
    } else {
      DeprecationNoticeType.OS_DEPRECATION
    }
  }
```

File: domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt
Comment (adhiamboperes) [inline domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt]:
```suggestion
  private fun computeAppStartupMode(onboardingState: OnboardingState): StartupMode {
    var startupMode: StartupMode = StartupMode.STARTUP_MODE_UNSPECIFIED
    deprecationController.getDeprecationResponse()
      .transform("GET_DEPRECATION_RESPONSE_PROVIDER_ID") { deprecationMode ->
        when (deprecationMode.deprecationNoticeType) {
          DeprecationNoticeType.OS_DEPRECATION -> {
            startupMode = StartupMode.OS_IS_DEPRECATED
          }
          DeprecationNoticeType.APP_DEPRECATION -> {
            // StartupMode.OPTIONAL_UPDATE_AVAILABLE or forced update??
            startupMode = StartupMode.OPTIONAL_UPDATE_AVAILABLE
          }
          else -> {
            startupMode = when {
              hasAppExpired() -> {
                StartupMode.APP_IS_DEPRECATED
              }
              onboardingState.alreadyOnboardedApp -> {
                StartupMode.USER_IS_ONBOARDED
              }
              else -> {
                StartupMode.USER_NOT_YET_ONBOARDED
              }
            }
          }
        }
      }
    return startupMode
  }
```

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
@kkmurerwa, I have offered suggestions for the data providers. Please double check the logic as this is mainly just a demo.

**Comment by kkmurerwa** [comment]
@adhiamboperes I have made a lot of changes including taking consideration of previous responses when showing the app deprecation dialog. PTAL

File: app/src/main/java/org/oppia/android/app/administratorcontrols/appversion/AppVersionViewModel.kt
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/administratorcontrols/appversion/AppVersionViewModel.kt:10]: Are these imports being used? There are no coresponding changes in the file.

File: app/src/main/java/org/oppia/android/app/notice/ForcedAppDeprecationNoticeDialogFragmentPresenter.kt
Comment (adhiamboperes) [inline app/src/main/java/org/oppia/android/app/notice/ForcedAppDeprecationNoticeDialogFragmentPresenter.kt]: Is this currently being used?

File: app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AppVersionActivityTest.kt
Comment (adhiamboperes) [inline app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AppVersionActivityTest.kt:93]: Please verify that these are needed

File: domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt
Comment (adhiamboperes) [reply domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt]: This LGTM in latest iteration.

File: domain/src/test/java/org/oppia/android/domain/onboarding/AppStartupStateControllerTest.kt
Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/onboarding/AppStartupStateControllerTest.kt]: I felt this would be more descriptive of the test conditions.
```suggestion
  fun testController_appAndOsDeprecationEnabled_initialLaunch_startupModeIsUserNotOnboarded() {
```

File: domain/src/test/java/org/oppia/android/domain/onboarding/AppStartupStateControllerTest.kt
Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/onboarding/AppStartupStateControllerTest.kt]: Maybe replace with what the correct expected `CorrectStartupMode` is, in case this changes in the future.

Comment (adhiamboperes) [reply domain/src/test/java/org/oppia/android/domain/onboarding/AppStartupStateControllerTest.kt]: Ditto other test names.

File: domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt
Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt]: I think this whole test could be simplified to:

```kotlin
fun testController_initialAppLaunch_returnsDefaultDeprecationResponseDatabase(){
setUpDefaultTestApplicationComponent()
val deprecationDataProvider = deprecationController.getDeprecationDatabase()
 val deprecationResponseDatabase = monitorFactory
      .waitForNextSuccessfulResult(deprecationDataProvider)
assertThat(deprecationResponseDatabase).isEqualTo(DeprecationResponseDatabase.getDefaultInstance())
}
```

This assertion is the same as splitting into two per your assertions.

File: domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt
Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt]: DRY: please extract 
```kotlin
 appDeprecationResponse = DeprecationResponse.newBuilder().apply {
        deprecatedVersion = 5
        deprecationNoticeType = DeprecationNoticeType.APP_DEPRECATION
      }.build()
``` 

to a variable.

Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt]:
```suggestion
  fun testController_previousResponseSaved_providesUpdatedDeprecationResponse() {
```

File: domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt
Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt]: The test names typically follow a format: `action_conditionOne_conditionTwo_expectedOutcome`
```suggestion
  fun testController_userNotOnboarded_returnsUserNotOnboardedStartUpMode() {
```

Comment (adhiamboperes) [reply domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt]: Please review other test names as well.

File: domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt
Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt]: Please remove this assertion.

File: domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt
Comment (adhiamboperes) [inline domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt]: We are not interested in this at this point. Ditto below.

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
Thanks @kkmurerwa! left some comments especially about test formats. Please see inline.

File: app/src/main/java/org/oppia/android/app/notice/ForcedAppDeprecationNoticeDialogFragmentPresenter.kt
Comment (kkmurerwa) [reply app/src/main/java/org/oppia/android/app/notice/ForcedAppDeprecationNoticeDialogFragmentPresenter.kt]: No it isn't. I have removed it.

File: app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AppVersionActivityTest.kt
Comment (kkmurerwa) [reply app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AppVersionActivityTest.kt:93]: They are needed. They were already there, I only changed the package name of the extensions file and these were refactored alongside that change

File: domain/src/test/java/org/oppia/android/domain/onboarding/AppStartupStateControllerTest.kt
Comment (kkmurerwa) [reply domain/src/test/java/org/oppia/android/domain/onboarding/AppStartupStateControllerTest.kt]: Done. I have also modified similar tests.

File: domain/src/test/java/org/oppia/android/domain/onboarding/AppStartupStateControllerTest.kt
Comment (kkmurerwa) [reply domain/src/test/java/org/oppia/android/domain/onboarding/AppStartupStateControllerTest.kt]: I have renamed the name for this and the other tests to meet this requirement

File: domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt
Comment (kkmurerwa) [reply domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt]: It doesn't seem to work since the two classes have different memory addresses and the assertion fails since it expects them to be the same object.

File: domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt
Comment (kkmurerwa) [reply domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt]: Done

Comment (kkmurerwa) [reply domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt]: I have renamed this and other tests

File: domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt
Comment (kkmurerwa) [reply domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt]: Done

File: domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt
Comment (kkmurerwa) [reply domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt]: I have removed the assertion and the other ones similar to it.

File: domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt
Comment (kkmurerwa) [reply domain/src/test/java/org/oppia/android/domain/onboarding/DeprecationControllerTest.kt]: Done

**Comment by kkmurerwa** [comment]
@adhiamboperes I have made the suggested changes. PTAL

**Review by adhiamboperes — APPROVED** [review]
Thanks @kkmurerwa! This LGTM. 

@BenHenning, PTAL for codeowners.

**Review by BenHenning — COMMENTED** [review]
As far as I can tell, this doesn't seem to require any review from me. I did take a quick scan and didn't notice anything to note, so going ahead and skipping review.

Deferring to you @adhiamboperes for merging in case I missed something.

**Comment by adhiamboperes** [comment]
Merging this since it LGTM.

**Pull request merged**
